File: old/webapp/src/main/scala/com/graphbrain/webapp/VisualGraph.scala
package com.graphbrain.webapp

import scala.math
import com.graphbrain.db._
import com.graphbrain.utils.JSONGen
import scala.collection.JavaConversions._
import scala.collection.mutable.Set

object VisualGraph {
  val MAX_SNODES = 15


  def generate(rootId: String, store: Graph, user: UserNode, edgeType: String = "", relPos: Integer = -1) = {
    val userId = if (user != null) user.id else ""
    
    // get neighboring edges
    val hyperEdges = store.edges(rootId, userId).filter(_.isPositive)

    // map hyperedges to visual edges
    val visualEdges = hyperEdges.map(hyper2edge(_, rootId)).filter(_ != null)

    // group nodes by edge type
    val edgeNodeMap = generateEdgeNodeMap(visualEdges, rootId)

    // truncate edge node map
    val truncatedEdgeNodeMap = truncateEdgeNodeMap(edgeNodeMap, MAX_SNODES)

    // full relations list
    val allRelations = edgeNodeMap.keys.map(x => Map("rel" -> x._1, "pos" -> x._2,
      "label" -> linkLabel(x._1), "snode" -> snodeId(x._1, x._2)))

    // create map with all information for supernodes
    val snodeMap = generateSnodeMap(truncatedEdgeNodeMap, store, rootId, user)

    // contexts
    val contexts = Nil

    // create reply structure with all the information needed for rendering
    val reply = Map("user" -> userId, "root" -> node2map(rootId, "", store, rootId, user),
      "snodes" -> snodeMap, "allrelations" -> allRelations, "context" -> "", "contexts" -> contexts)

    //WebServer.graph.clear()

    // generate json reply
    JSONGen.json(reply)
  }

  private def hyper2edge(edge: Edge, rootId: String): SimpleEdge = {
    if (edge.getParticipantIds.length > 2) {
      if (edge.getEdgeType == "r/1/instance_of~owned_by") {
        if (edge.getParticipantIds()(0) == rootId) {
          new SimpleEdge("r/1/has", edge.getParticipantIds()(2), rootId, edge)
        }
        else if (edge.getParticipantIds()(2) == rootId) {
          new SimpleEdge("r/1/has", rootId, edge.getParticipantIds()(0), edge)
        }
        else {
          null
        }
      }
      if (edge.getEdgeType == "r/1/has~of_type") {
        val edgeType = "has " + ID.lastPart(edge.getParticipantIds()(2))
        new SimpleEdge(edgeType, edge.getParticipantIds()(0), edge.getParticipantIds()(1), edge)
      }
      else {
        val parts = edge.getEdgeType.split("~").toList
        val edgeType = parts.head + " " + ID.lastPart(edge.getParticipantIds()(1)) + " " + parts.tail.reduceLeft(_ + " " + _)
        //val edgeType = edge.edgeType.replaceAll("~", " .. ") + " .. "
        new SimpleEdge(edgeType, edge.getParticipantIds()(0), edge.getParticipantIds()(2), edge)
      }
    }
    else {
      new SimpleEdge(edge)
    }
  }

  private def generateEdgeNodeMap(edges: Set[SimpleEdge], rootId: String) = {
    edges.map(
      e => List(e.getId1 -> 0, e.getId2 -> 1)
        .map(x => (e.getEdgeType, x._2, x._1, e.getParent.toString))
    ).flatten
      .filter(x => x._3 != rootId)
      .groupBy(x => (x._1, x._2))
      .mapValues(x => x.map(y => (y._3, y._4)))
  }

  private def truncateEdgeNodeMap(edgeNodeMap: Map[(String, Int), Set[(String, String)]], maxSize: Int) = {
    edgeNodeMap.zipWithIndex.filter(x=> x._2 < maxSize).map(x => x._1)
  }

  private def node2map(nodeId: String, nodeEdge: String, store: Graph, rootId: String, user: UserNode) = {
    val vtype = VertexType.getType(nodeId)
    val node = if ((nodeId != rootId) && (vtype == VertexType.Entity)) {
      new EntityNode(nodeId)
    }
    else{
      try {
        store.get(nodeId)
      }
      catch {
        case _: Throwable => null
      }
    }

    node match {
      case tn: EntityNode => Map("id" -> tn.id, "type" -> "text", "text" -> tn.text, "edge" -> nodeEdge)
      case un: URLNode => {
        val title = if (un.getTitle == "") un.getUrl else un.getTitle
        Map("id" -> un.id, "type" -> "url", "text" -> title, "url" -> un.getUrl, "icon" -> un.getIcon, "edge" -> nodeEdge)
      }
      case un: UserNode => Map("id" -> un.id, "type" -> "user", "text" -> un.getName, "edge" -> nodeEdge)
      case null => ""
      case _ => Map("id" -> node.id, "type" -> "text", "text" -> node.id, "edge" -> nodeEdge)
    }
  }

  private def snodeId(edgeType: String, pos: Integer) = edgeType.replaceAll("/", "_").replaceAll(" ", "_").replaceAll("\\.", "_") + "_" + pos

  private def generateSnode(pair: ((String, Int), Set[(String, String)]), store: Graph, rootId: String, user: UserNode) = {
    val id = snodeId(pair._1._1, pair._1._2)
    val label = linkLabel(pair._1._1)
    val color = linkColor(label)
    val nodes = pair._2.map(x => node2map(x._1, x._2, store, rootId, user))

    val data = Map("nodes" -> nodes, "etype" -> pair._1._1, "rpos" -> pair._1._2, "label" -> label, "color" -> color)

    id -> data
  }

  private def generateSnodeMap(edgeNodeMap: Map[(String, Int), Set[(String, String)]], store: Graph, rootId: String, user: UserNode) = {
    edgeNodeMap.map(x => generateSnode(x, store, rootId, user))
  }

  private def linkColor(label: String) = {
    val index = math.abs(label.hashCode) % Colors.colors.length
    Colors.colors(index)
  }

  private def fixLabel(label: String) = if (EdgeLabelTable.table.contains(label)) EdgeLabelTable.table(label) else label

  private def linkLabel(edgeType: String): String = {
    if (edgeType == "")
      return ""
    val lastPart = ID.parts(edgeType).last
    fixLabel(lastPart.replace("_", " "))
  }

  /*
  def main(args: Array[String]) {
    val store = new VertexStore with SimpleCaching with UserOps with UserManagement
      
    println(VisualGraph.generate("1/eraserhead", store, null))

    sys.exit(0)
  }*/
}

File: web/src/main/java/com/graphbrain/web/VisualGraph.java
package com.graphbrain.web;

import com.graphbrain.db.*;
import org.apache.commons.lang.builder.HashCodeBuilder;
import org.json.JSONObject;

import java.util.*;

public class VisualGraph {

    public static final int MAX_SNODES = 15;

    private static class RelPos {
        public String rel;
        public int pos;

        public RelPos(String rel, int pos) {
            this.rel = rel;
            this.pos = pos;
        }

        public String snodeId() {
            return rel.replaceAll("/", "_").replaceAll(" ", "_").replaceAll("\\.", "_").replaceAll("\\+", "_") + "_" + pos;
        }

        @Override
        public int hashCode() {
            return new HashCodeBuilder(17, 31).append(rel).append(pos).toHashCode();
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == null)
                return false;
            if (obj == this)
                return true;
            if (!(obj instanceof RelPos))
                return false;

            RelPos rp = (RelPos)obj;
            return (pos == rp.pos) && (rel.equals(rp.rel));
        }
    }

    public static String generate(String rootId, UserNode user) {
        String userId = "";
        if (user != null)
            userId = user.id;

        // get neighboring edges
        Set<Edge> hyperEdges = WebServer.graph.edges(rootId, userId);

        // map hyperedges to visual edges
        Set<SimpleEdge> visualEdges = new HashSet<SimpleEdge>();
        for (Edge e: hyperEdges) {
            if (e.isPositive()) {
                SimpleEdge se = hyper2edge(e, rootId);
                if (se != null) {
                    visualEdges.add(se);
                }
            }
        }

        // group nodes by edge type
        Map<RelPos, Set<SimpleEdge>> edgeNodeMap = generateEdgeNodeMap(visualEdges, rootId);

        // full relations list
        List<Map<String, String>> allRelations = new LinkedList<Map<String, String>>();
        for (RelPos rp : edgeNodeMap.keySet()) {
            Map<String, String> r = new HashMap<String, String>();
            r.put("rel", rp.rel);
            r.put("pos", "" + rp.pos);
            r.put("label", linkLabel(rp.rel));
            r.put("snode", rp.snodeId());
            allRelations.add(r);
        }

        // create map with all information for supernodes
        Map<String, Map<String, Object>> snodeMap = generateSnodeMap(edgeNodeMap, rootId);

        // create reply structure with all the information needed for rendering
        JSONObject json = new JSONObject();
        json.put("user", userId);
        json.put("root", node2map(rootId, "", rootId));
        json.put("snodes", snodeMap);
        json.put("allrelations", allRelations);

        return json.toString();
    }

    private static SimpleEdge hyper2edge(Edge edge, String rootId) {
        if (edge.getParticipantIds().length > 2) {
            if (edge.getEdgeType().equals("r/1/instance_of~owned_by")) {
                if (edge.getParticipantIds()[0].equals(rootId)) {
                    return new SimpleEdge("r/1/has", edge.getParticipantIds()[2], rootId, edge);
                }
                else if (edge.getParticipantIds()[2].equals(rootId)) {
                    return new SimpleEdge("r/1/has", rootId, edge.getParticipantIds()[0], edge);
                }
                else {
                    return null;
                }
            }
            if (edge.getEdgeType().equals("r/1/has~of_type")) {
                String edgeType = "has " + ID.lastPart(edge.getParticipantIds()[2]);
                return new SimpleEdge(edgeType, edge.getParticipantIds()[0], edge.getParticipantIds()[1], edge);
            }
            else {
                String[] parts = edge.getEdgeType().split("~");
                String edgeType = parts[0];
                edgeType += " ";
                edgeType += ID.lastPart(edge.getParticipantIds()[1]);
                for (int i = 1; i < parts.length; i++) {
                    edgeType += " ";
                    edgeType += parts[i];
                }

                return new SimpleEdge(edgeType, edge.getParticipantIds()[0], edge.getParticipantIds()[2], edge);
            }
        }
        else {
            return new SimpleEdge(edge);
        }
    }

    private static void addToEdgeNodeMap(Map<RelPos, Set<SimpleEdge>> enMap, RelPos key, SimpleEdge e, String rootId) {
        if (key.pos == 0) {
            if (e.getId1().equals(rootId))
                return;
        }
        if (key.pos == 1) {
            if (e.getId2().equals(rootId))
                return;
        }

        if (enMap.containsKey(key)) {
            enMap.get(key).add(e);
        }
        else {
            Set<SimpleEdge> set = new HashSet<>();
            set.add(e);
            enMap.put(key, set);
        }
    }

    private static Map<RelPos, Set<SimpleEdge>> generateEdgeNodeMap(Set<SimpleEdge> edges, String rootId) {
        Map<RelPos, Set<SimpleEdge>> enMap = new HashMap<>();

        int count = 0;
        for (SimpleEdge e : edges) {
            addToEdgeNodeMap(enMap, new RelPos(e.getEdgeType(), 0), e, rootId);
            addToEdgeNodeMap(enMap, new RelPos(e.getEdgeType(), 1), e, rootId);

            count++;
            if (count > MAX_SNODES) {
                return enMap;
            }
        }

        return enMap;
    }

    private static Map<String, String> node2map(String nodeId, String nodeEdge, String rootId) {
        VertexType vtype = VertexType.getType(nodeId);
        Vertex node;

        if ((nodeId.equals(rootId)) && (vtype == VertexType.Entity)) {
            node = new EntityNode(nodeId);
        }
        else {
            node = WebServer.graph.get(nodeId);
        }

        Map<String, String> map = new HashMap<>();

        switch(node.type()) {
            case Entity:
                EntityNode en = (EntityNode)node;
                map.put("id", en.id);
                map.put("type", "text");
                map.put("text", WebServer.graph.description(en));
                map.put("edge", nodeEdge);
                return map;
            case URL:
                URLNode un;
                if (node instanceof URLNode)
                    un = (URLNode)node;
                else
                    return map;
                String title;
                if (un.getTitle().equals(""))
                    title = un.getUrl();
                else
                    title = un.getTitle();
                map.put("id", un.id);
                map.put("type", "url");
                map.put("text", title);
                map.put("url", un.getUrl());
                map.put("icon", un.getIcon());
                map.put("edge", nodeEdge);
                return map;
            case User:
                UserNode us;
                if (node instanceof UserNode)
                    us = (UserNode)node;
                else
                    return map;
                map.put("id", us.id);
                map.put("type", "user");
                map.put("text", us.getName());
                map.put("edge", nodeEdge);
                return map;
            default:
                map.put("id", node.id);
                map.put("type", "text");
                map.put("text", node.id);
                map.put("edge", nodeEdge);
                return map;
        }
    }

    private static Map<String, Object> generateSnode(RelPos rp, Set<SimpleEdge> sedges, String rootId) {
        String label = linkLabel(rp.rel);
        String color = linkColor(label);
        List<Map<String, String>> nodes = new LinkedList<>();

        for (SimpleEdge se : sedges) {
            String nodeId;
            if (rp.pos == 0) {
                nodeId = se.getId1();
            }
            else {
                nodeId = se.getId2();
            }

            nodes.add(node2map(nodeId, se.getParent().id, rootId));
        }

        Map<String, Object> data = new HashMap<>();
        data.put("nodes", nodes);
        data.put("etype", rp.rel);
        data.put("rpos", rp.pos);
        data.put("label", label);
        data.put("color", color);

        return data;
    }

    private static Map<String, Map<String, Object>> generateSnodeMap(Map<RelPos, Set<SimpleEdge>> edgeNodeMap, String rootId) {
        Map<String, Map<String, Object>> snodeMap = new HashMap<String, Map<String, Object>>();

        for (RelPos rp : edgeNodeMap.keySet()) {
            snodeMap.put(rp.snodeId(), generateSnode(rp, edgeNodeMap.get(rp), rootId));
        }

        return snodeMap;
    }

    private static String linkColor(String label) {
        int index = Math.abs(label.hashCode()) % Colors.colors.length;
        return Colors.colors[index];
    }

    private static String fixLabel(String label) {
        if (EdgeLabelTable.lt.containsKey(label))
            return EdgeLabelTable.lt.get(label);
        else
            return label;
    }

    private static String linkLabel(String edgeType) {
        if (edgeType.isEmpty())
            return "";

        return fixLabel(EdgeType.label(edgeType));
    }
}
